### PR TITLE
add tests for overflow, fix bugs encountered

### PR DESCRIPTION
**Pull request by WalterBright**
Added some tests to test uncovered overflow logic for static arrays, and found bugs. Still not convinced that test coverage checking is worthless!

### CODE REVIEW

File: src/toobj.c [hunk 1]
```diff
@@ -949,7 +949,17 @@ void toObjFile(Dsymbol *ds, bool multiobj)
...
 
             Symbol *s = toSymbol(vd);
-            unsigned sz = vd->type->size();
+            d_uns64 sz64 = vd->type->size(vd->loc);
+            if (sz64 == SIZE_INVALID)
+            {
+                vd->error("size overflow");
+                return;
+            }
+            if (sz64 >= 0x1000000)
```
Comment (ZombineDev) [inline src/toobj.c]: Please refactor the magic number into a proper constant and explain in the message below why the size is limited to `0x1000000` (i.e. why is it a problem).

Comment (WalterBright) [inline src/toobj.c]: I'm not enamored with using a named constant for a single use magic number, because it adds a pointless level of indirection for the cognition of the user.

Comment (WalterBright) [inline src/toobj.c]: I added a comment

File: src/toobj.c [hunk 2]
```diff
@@ -949,7 +949,17 @@ void toObjFile(Dsymbol *ds, bool multiobj)
...
-            unsigned sz = vd->type->size();
+            d_uns64 sz64 = vd->type->size(vd->loc);
+            if (sz64 == SIZE_INVALID)
+            {
+                vd->error("size overflow");
+                return;
+            }
+            if (sz64 >= 0x1000000)  // there has to be some 'reasonable' limit on the size
+            {
+                vd->error("size of x%llx exceeds max allowed size 0x100_0000", sz64);
```
Comment (ZombineDev) [inline src/toobj.c]: Thanks, that's a little better

**Comment by ZombineDev** [comment]
LGTM

**Comment by wilzbach** [comment]
OT: I start to believe that the deadlocking on OS X 32-bit is systematic (I restarted the build here too), but afaik is OS X 64-bit only since a long time, so is it even worthwhile to run the 32-bit test?
